### PR TITLE
engine: in-place updates for files from the base image

### PR DESCRIPTION
**Pull request by nicks**
Hello @maiamcc,

Please review the following commits I made in branch nicks/liveupdategraph2:

ed079fd30eabb1ac8d4106d6f564eba0a9793577 (2019-04-28 23:33:29 -0400)
engine: in-place updates for files from the base image
Note that this commit doesn't fix the underlying bug yet, because
we still need to update the tiltfile to allow these kinds of live-updates

### CODE REVIEW

File: internal/engine/build_and_deployer_live_update_test.go [hunk 1]
```diff
@@ -463,3 +468,61 @@ func TestLiveUpdateSomeFilesMatchSyncSomeDontFallsBack(t *testing.T) {
 	}
 	runTestCase(t, f, tCase)
 }
+
+func TestLiveUpdateInFirstImageOfImageDependency(t *testing.T) {
```
Comment (maiamcc) [inline internal/engine/build_and_deployer_live_update_test.go]: nice! could you also test that, without an explicit sync step for sancho-base, a changed file in sancho-base prompts a fallback to full build?

File: internal/engine/live_update_state_tree.go [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+	"github.com/windmilleng/tilt/internal/store"
+)
+
+// A helper data structure that represents a live-update image and
+// the files changed in all of its dependencies.
+type liveUpdateStateTree struct {
+	iTarget            model.ImageTarget
+	filesChanged       []string
+	iTargetState       store.BuildState
+	needsImageBuildIDs []model.TargetID
```
Comment (maiamcc) [inline internal/engine/live_update_state_tree.go]: see naming note

File: internal/engine/build_and_deployer_live_update_test.go (hunk 1 above)
Comment (nicks) [reply internal/engine/build_and_deployer_live_update_test.go]: i like that! done

File: internal/engine/extractors.go [hunk 3]
```diff
@@ -21,21 +25,33 @@ func extractImageAndK8sTargets(specs []model.TargetSpec) (iTargets []model.Image
 }
 
 // Extract image targets iff they can be updated in-place in a container.
```
[the remark this answers is not in the document]
Comment (nicks) [reply internal/engine/extractors.go]: done

File: internal/engine/extractors.go [hunk 4]
```diff
@@ -21,21 +25,33 @@ func extractImageAndK8sTargets(specs []model.TargetSpec) (iTargets []model.Image
...
+	deployedImages := g.DeployedImages()
+	for _, iTarget := range deployedImages {
 		state := stateSet[iTarget.ID()]
 		if state.IsEmpty() {
 			return nil, SilentRedirectToNextBuilderf("In-place build does not support initial deploy")
 		}
 
-		// If this image doesn't need to be built at all, we can skip it.
-		if !state.NeedsImageBuild() {
+		needsImageBuildIDs, err := needsImageBuildTree(g, iTarget, stateSet)
```
[the remark this answers is not in the document]
Comment (nicks) [reply internal/engine/extractors.go]: i changed this to hasFilesChanged, which is a better representation of what we want

File: internal/engine/extractors.go [hunk 5]
```diff
@@ -82,3 +110,35 @@ func isImageDeployedToDC(iTarget model.ImageTarget, dcTarget model.DockerCompose
 	}
 	return false
 }
+
+// Given a target, return all the target IDs in its tree of dependencies that
+// need a new image build.
+func needsImageBuildTree(g model.TargetGraph, target model.TargetSpec, stateSet store.BuildStateSet) ([]model.TargetID, error) {
```
[the remark this answers is not in the document]
Comment (nicks) [reply internal/engine/extractors.go]: ditto

File: internal/engine/extractors.go [hunk 6]
```diff
@@ -82,3 +110,35 @@ func isImageDeployedToDC(iTarget model.ImageTarget, dcTarget model.DockerCompose
...
+	result := []model.TargetID{}
+	err := g.VisitTree(target, func(current model.TargetSpec) error {
+		state := stateSet[current.ID()]
+		if state.NeedsImageBuild() {
```
[the remark this answers is not in the document]
Comment (nicks) [reply internal/engine/extractors.go]: ditto

File: internal/engine/live_update_state_tree.go (hunk 2 above)
Comment (nicks) [reply internal/engine/live_update_state_tree.go]: ditto

File: internal/engine/live_update_state_tree.go [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+func (t liveUpdateStateTree) createResultSet() store.BuildResultSet {
+	iTargetID := t.iTarget.ID()
+	state := t.iTargetState
+	res := state.LastResult.ShallowCloneForContainerUpdate(state.FilesChangedSet)
+	res.ContainerID = state.DeployInfo.ContainerID
+
+	resultSet := store.BuildResultSet{}
+	resultSet[iTargetID] = res
+
+	// Invalidate all the image builds for images we depend on.
```
[the remark this answers is not in the document]
Comment (nicks) [reply internal/engine/live_update_state_tree.go]: If we live the image ID as-is, the image builder will think no files have changed and that it doesn't need to rebuild anything.

**Comment by nicks** [comment]
ya, the naming issue was a sign that we were checking the wrong thing. we really want to check filesChanged, not NeedsImageBuild

File: internal/engine/extractors.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (maiamcc) [reply internal/engine/extractors.go]: sgtm

**Pull request merged**